### DISCUSSION TITLE
Fan Out Jobs

### DISCUSSION THREAD

**Discussion by fotoflo** [quirrel-dev/quirrel#1127] [Q&A] [answered]
Hi everyone, I was checking out the Quirrel noticed that there is mention of fanout jobs on the home page, but I couldn't find any documentation about them. 

I was able to get it a cron job working on my local host, but of course that doesn't create fan out jobs.

Does this automatically fan out on serverless functions or do i use the exclusive peremeter or something? 

Any insights or resources would be greatly appreciated. Thank you!

**Comment by Skn0tt** [comment, marked as answer]
Hi @fotoflo! Sorry for the late reply, I was on PTO :D

To implement a fan-out job, you can create a `Queue` and then call [`.enqueue`](https://docs.quirrel.dev/api/queue#enqueue) or [`.enqueueMany`](https://docs.quirrel.dev/api/queue#enqueuemany) on it, to create jobs that will be processed in other serverless functions.

**Reply by fotoflo** [reply]
> Hi @fotoflo! Sorry for the late reply, I was on PTO :D
> 
> 
> 
> To implement a fan-out job, you can create a `Queue` and then call [`.enqueue`](https://docs.quirrel.dev/api/queue#enqueue) or [`.enqueueMany`](https://docs.quirrel.dev/api/queue#enqueuemany) on it, to create jobs that will be processed in other serverless functions.

Great! Thanks @Skn0tt! 

Was hoping that was the case. It's all coded up and working on local. Will be deploying soon so can't wait to see it in action.

I am using the 'exclusive' flag on local to ensure my task runs alone.  (I'm running a lighthouse report and they generally should be run exclusively)

Assuming this is also the right thing to do for fan out jobs?

Thanks again for your response

**Reply by Skn0tt** [reply]
> Assuming this is also the right thing to do for fan out jobs?

Depends on what you're trying to achieve. Using `exclusive` will mean that there's only one job run at the same time, now matter how many jobs there are. For something long-running like lighthouse reports, that can become a bottleneck. Could you elaborate as to why you're intending for them to run exclusively?

**Reply by Skn0tt** [reply]
> Will be deploying soon so can't wait to see it in action.

Just to make sure you don't hit a wall, because people have been bitten by this before: There's no hosted offering anymore, and you'll need to stand up your own Quirrel instance.

**Reply by fotoflo** [reply]
Thanks yeah I see that. I was thinking to setup a upstash redis instance and then it looks like I just drop the api keys/database uri in vercel, but not sure about how to use the quirrel ci command... will investigate

**Reply by fotoflo** [reply]
> > Assuming this is also the right thing to do for fan out jobs?
> 
> 
> 
> Depends on what you're trying to achieve. Using `exclusive` will mean that there's only one job run at the same time, now matter how many jobs there are. For something long-running like lighthouse reports, that can become a bottleneck. Could you elaborate as to why you're intending for them to run exclusively?

This is how I have to do it on local... will the jobs fan out if I use exclusive in vercel?

**Reply by Skn0tt** [reply]
> will the jobs fan out if I use exclusive in vercel?

No, they won't. The `exclusive` option is for operations that *strictly* require sequential processing, like some usecases in payment + order fulfilment.

> I was thinking to setup a upstash redis instance

Let me know if that works well - my understanding is that Upstash is a custom redis implementation, and i've never tested it with Quirrel. You'll still have to host the Quirrel docker image somewhere, though.
